### DISCUSSION TITLE
libminizip

### DISCUSSION THREAD

**Discussion by MarkCallow** [assimp/assimp#4442] [Q&A]
I am building the static library version of v5.2.2 on macOS with only the importers. The compiled library has references to various `_unz*` functions which are defined in libminizip.a. Even if I set ASSIMP_BUILD_ZLIB the compiled library still wants the libminizip functions.

I have a libminizip from when I built libassimp a long time ago. I think at that time it must have been included in the assimp repo. If I use that and I'm on an x86_64 mac then my app builds and works. The problem is I need to build a universal binary of `libminizip.a`. I have 2 questions:

1. How can I get assimp to use the libzlibstatic.a that the project builds instead of libminizip? I saw nothing in the CMake config options.
2. Alternatively where is the source for libminizip so I can build a universal binary version. I've found many on the web, some very old, some newer and larger. Which is the one that will work for libassimp?

[This](https://github.com/domoticz/minizip) minizip works. I'd still like to know the answer to my first question.

Build.md needs info about the zip libraries.

**Comment by kimkulling** [comment]
Hm, do you know where the minizip stuff will be used? I am wondering why we still need this dependency. I guess I have intorduced this years ago for whatever reason.

I need to check if we can offer you a way to define your own minizip version.

**Comment by kimkulling** [comment]
The only workaround I found is using the hunter-option to download the minzip lib automatically. At this moment there is no other option I guess.

**Comment by MarkCallow** [comment]
All I know is I built the static library `libassimp.a` without exporters, tests or tools and when I linked my application with it I got undefined references to the _unz functions. They aren't coming from my app. I also discovered that libminizip references libz.a, by removing it from my link command - all the references to it came from _unz* functions. So there doesn't seem any advantage in using minizip.

**Comment by kimkulling** [comment]
Could you provide the list with unresolved symbols? And could you provide your used compiler, please?

**Reply by MarkCallow** [reply]
Sorry for the delay. Here is the list along with the names of the .o files from libassimp.a that are referencing them:

```
Undefined symbols for architecture x86_64:
  "_unzClose", referenced from:
      Assimp::IFCImporter::InternReadFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, aiScene*, Assimp::IOSystem*) in libassimp.a(IFCLoader.o)
      Assimp::ZipArchiveIOSystem::Implement::~Implement() in libassimp.a(ZipArchiveIOSystem.o)
      Assimp::ZipArchiveIOSystem::Implement::~Implement() in libassimp.a(ZipArchiveIOSystem.o)
      Assimp::ZipArchiveIOSystem::~ZipArchiveIOSystem() in libassimp.a(ZipArchiveIOSystem.o)
      Assimp::ZipArchiveIOSystem::isZipArchive(Assimp::IOSystem*, char const*) in libassimp.a(ZipArchiveIOSystem.o)
  "_unzCloseCurrentFile", referenced from:
      Assimp::IFCImporter::InternReadFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, aiScene*, Assimp::IOSystem*) in libassimp.a(IFCLoader.o)
  "_unzGetCurrentFileInfo", referenced from:
      Assimp::IFCImporter::InternReadFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, aiScene*, Assimp::IOSystem*) in libassimp.a(IFCLoader.o)
      Assimp::ZipArchiveIOSystem::Implement::MapArchive() in libassimp.a(ZipArchiveIOSystem.o)
  "_unzGetFilePos", referenced from:
      Assimp::ZipFileInfo::ZipFileInfo(void*, unsigned long) in libassimp.a(ZipArchiveIOSystem.o)
      Assimp::ZipFileInfo::ZipFileInfo(void*, unsigned long) in libassimp.a(ZipArchiveIOSystem.o)
      Assimp::ZipArchiveIOSystem::Implement::MapArchive() in libassimp.a(ZipArchiveIOSystem.o)
  "_unzGoToFilePos", referenced from:
      Assimp::ZipFileInfo::Extract(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, void*) const in libassimp.a(ZipArchiveIOSystem.o)
  "_unzGoToFirstFile", referenced from:
      Assimp::IFCImporter::InternReadFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, aiScene*, Assimp::IOSystem*) in libassimp.a(IFCLoader.o)
      Assimp::ZipArchiveIOSystem::Implement::MapArchive() in libassimp.a(ZipArchiveIOSystem.o)
  "_unzGoToNextFile", referenced from:
      Assimp::IFCImporter::InternReadFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, aiScene*, Assimp::IOSystem*) in libassimp.a(IFCLoader.o)
      Assimp::ZipArchiveIOSystem::Implement::MapArchive() in libassimp.a(ZipArchiveIOSystem.o)
  "_unzOpen", referenced from:
      Assimp::IFCImporter::InternReadFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, aiScene*, Assimp::IOSystem*) in libassimp.a(IFCLoader.o)
  "_unzOpen2", referenced from:
      Assimp::ZipArchiveIOSystem::Implement::Implement(Assimp::IOSystem*, char const*, char const*) in libassimp.a(ZipArchiveIOSystem.o)
      Assimp::ZipArchiveIOSystem::Implement::Implement(Assimp::IOSystem*, char const*, char const*) in libassimp.a(ZipArchiveIOSystem.o)
      Assimp::ZipArchiveIOSystem::ZipArchiveIOSystem(Assimp::IOSystem*, char const*, char const*) in libassimp.a(ZipArchiveIOSystem.o)
      Assimp::ZipArchiveIOSystem::ZipArchiveIOSystem(Assimp::IOSystem*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const*) in libassimp.a(ZipArchiveIOSystem.o)
      Assimp::ZipArchiveIOSystem::isZipArchive(Assimp::IOSystem*, char const*) in libassimp.a(ZipArchiveIOSystem.o)
  "_unzOpenCurrentFile", referenced from:
      Assimp::IFCImporter::InternReadFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, aiScene*, Assimp::IOSystem*) in libassimp.a(IFCLoader.o)
      Assimp::ZipFileInfo::Extract(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, void*) const in libassimp.a(ZipArchiveIOSystem.o)
  "_unzReadCurrentFile", referenced from:
      Assimp::IFCImporter::InternReadFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, aiScene*, Assimp::IOSystem*) in libassimp.a(IFCLoader.o)
      Assimp::ZipFileInfo::Extract(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, void*) const in libassimp.a(ZipArchiveIOSystem.o)
```

**Comment by sandervalcke** [comment]
Ran into the same issue. FYI minizip is something completely separate from zlib, it provides _zip_ functionality, which (confusingly) zlib doesn't provide, zlib provides the algorithm for gzip compression. Assimp needs minizip, which it (no longer?) builds. I ended up adding a small project to build minizip statically, similar to [this](https://stackoverflow.com/a/39942944/678022), and then link my project that consumes assimp (statically) against minizip as well.
